### ISSUE TITLE
go build

### ISSUE THREAD

**Issue by r03** [go-lang-plugin-org/go-lang-idea-plugin#1636]
I'm new to IntelliJ/WebStorm

How can I do a simple 'go build' in the current working dir?
I tried to leave the file empty in a go application config, but that isn't possible.

**Comment by dlsniper** [comment]
Hi, you can create a run configuration that's called Go Application and select the package option. 

Please see the screenshot below.

![snapshot94](https://cloud.githubusercontent.com/assets/607868/7746781/73d07280-ffb5-11e4-87e0-aa108060db43.png)

**Comment by r03** [comment]
That only works if the package is in the GOPATH
Since I can't leave package empty this also means I need to create a build configuration per package?

Is there a way to just run a command directly? (go build)

**Comment by dlsniper** [comment]
Please detail with examples what you want to achieve as it's not clear. Isn't the package in the GOPATH to begin with? How did you created the project in IDEA? I have a project called ```demo``` which is under GOPATH/src/github.com/dlsniper/demo and the root for project in IDEA is /home/florinp/golang/src/github.com/dlsniper/demo (with GOPATH=/home/florinp/golang)

**Comment by r03** [comment]
No, the package isn't in the GOPATH. It's not a library so it is not really needed.
But I will move the source to the GOPATH for now.

![image](https://cloud.githubusercontent.com/assets/10061147/7747721/9ee70fee-ffbd-11e4-8a69-3d10b1548902.png)

**Comment by dlsniper** [comment]
@r03 thanks for clarifying.

I'll reopen this, @zolotov what do you think about this kind of setup?
- the application is in ```/path/to/dir/of/the/app```
- the GOAPTH is in ```/path/go/gopath```

```go build``` does indeed seem to work ok without the layout of the directory containing a ```src/``` or anything else as long as there's a buildable source there.

**Comment by r03** [comment]
Also using gb is not possible: http://getgb.io/
That's ok since it is not an 'official' tool, but maybe you can make the build configuration more flexible?
Maybe allow a custom command or something?

**Comment by dlsniper** [comment]
We happily accept contributions :)

**Comment by zolotov** [comment]
> Maybe allow a custom command or something?

What's the point? Any custom command you can invoke in terminal (real one builtin one). The only reason why `Go Application` run configuration is needed – it passes all project configuration that IDEA knows (module/project-level GOPATH, GOROOT, etc.) to go tool.

As for gb support, +1 to @dlsniper

**Comment by r03** [comment]
The point is that I'm trying to get a better IDE experience. F5 is easier then typing 'go build' :-)
I was also hoping that I could click on a build error and then go to the correct line in the source.

**Comment by zolotov** [comment]
You can configure [external tool](https://www.jetbrains.com/idea/help/external-tools.html) and assign any shortcut to it. It's quite customizable.

Links:
https://github.com/go-lang-plugin-org/go-lang-idea-plugin/blob/master/CONTRIBUTING.md#getting-started-with-the-plugin-development

**Comment by r03** [comment]
Thanks, I'll give it a try.

**Comment by zolotov** [comment]
> I'll reopen this, @zolotov what do you think about this kind of setup?

> the application is in /path/to/dir/of/the/app
> the GOAPTH is in /path/go/gopath

We're definitely not going to support not recommended ways of building go applications. But building on directory makes sense despite of its location.

**Comment by pires** [comment]
+1 for gb integration.

**Comment by iampeter** [comment]
+1 for gb integration. any changes so far?

**Comment by dlsniper** [comment]
Nothing has changed so far, please read the above comments to understand why. If someone else is interested in adding support for this, then support can be offered to that person.

Finally, gb is not an official go tool, that's why this is not a priority (at least afaik) and support from the community would be very much appreciated.

Also, please contact Github to add a voting feature to issues as sending a +1 email is a really bad way to convey your needed (not to mention sending a mail to people who are not maintainers and might have been subscribed / commented to the issue). Alternatively, you can choose to subscribe to an issue to know when it's status changes.  

Thank you.

**Comment by pbennett** [comment]
Formal build support would indeed be nice (or at least more refined external tool usage)
External tools allows me to run say, go install for the current files directory but the ide is then oblivious about errors or allow me to navigate to them.  From what I can tell, I should be able to specify an output filter of $FILE$:$LINE$ but this doesn't match go build/install error syntax.
perhaps it doesn't like ./xxxx.go:55: some message ?
Coming from a C++ world, having an IDE that doesn't have a built-in concept of 'building' is a bit strange for me.

**Comment by zolotov** [comment]
> Formal build support would indeed be nice 

Formal build support is in here. As it was said in comments you need external tools for unrecommended scenario when your project is not under GOPATH. At the moment we're not going to make better experience for unrecommended project setup.

**Comment by rand99** [comment]
I just switched from getgb.io to govendor.
(after messing around to get gb working with the plugin adding paths and .. and ...)

govendor works fine with the go-lang-idea-plugin without configuring anything. 
I'm using go 1.6 and I have my project in the go-path src folder (symlink but works :)

@zolotov Just tell the people on the main page of the repo to use govendor and they will be happy.
https://github.com/kardianos/govendor
no more pain

**Comment by dlsniper** [comment]
@rand99 unfortunately your suggestion doesn't apply to many many many projects.
First, govendor is not the only tool out there, second, there are projects which are happily using go get only. There is support for vendor experiment (no more symlinks messing around, which is pain, so govendor doesn't fix anything) coming soon (should be in the next release).

Also your solution has nothing to do with the original request nor the problems around it. @zolotov should we close this issue meanwhile?

Finally I'd rather say the plugin should be impartial, at least for now, and keep continue supporting what the official Go supports.
